src/order/filter/filter_product.lean

```diff
@@ -0,0 +1,261 @@
+/-
+Copyright (c) 2019 Abhimanyu Pallavi Sudhir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Abhimanyu Pallavi Sudhir
+"Filterproducts" (ultraproducts on general filters), ultraproducts and 
+the construction of the hyperreals.
+-/
+
+import data.real.basic
+import order.filter.basic
+
+universes u v
+variables {α : Type u} (β : Type v) (φ : filter α)
+local attribute [instance] classical.prop_decidable
+
+namespace filter 
+
+/-- Two sequences are bigly equal iff the kernel of their difference is in φ -/
+def bigly_equal : setoid (α → β) := 
+⟨ λ a b, {n | a n = b n} ∈ φ.sets,
+  λ a, by simp only [eq_self_iff_true, (set.univ_def).symm, univ_sets], 
+  λ a b ab, by simpa only [eq_comm], 
+  λ a b c ab bc, sets_of_superset φ (inter_sets φ ab bc) 
+    (λ n (r : a n = b n ∧ b n = c n), eq.trans r.1 r.2)⟩
+
+/-- Ultraproduct, but on a general filter -/
+def filterprod := quotient (bigly_equal β φ)
+local notation `β*` := filterprod β φ
+
+namespace filter_product
+
+def seq_to_filterprod : (α → β) → β* := @quotient.mk' (α → β) (bigly_equal β φ)
+
+/-- Equivalence class containing the constant sequence of a term in β -/
+def to_filterprod : β → β* := function.comp (seq_to_filterprod β φ) (λ x n, x)
+
+instance : has_coe (α → β) β* :=  ⟨seq_to_filterprod β φ⟩
+
+instance coe_filterprod : has_coe β β* := ⟨to_filterprod β φ⟩
+
+section ultraproduct
+variable {U : is_ultrafilter φ} include U
+
+theorem to_filterprod_inj : function.injective (to_filterprod β φ) :=
```

Comment:
For the injectivity, you don't need phi to be a ultrafilter, it should just be nontrivial. Also, I would put this as an explicit argument instead of an implicit one.